graphics/src/text/paragraph.rs

```diff
@@ -0,0 +1,246 @@
+use crate::core;
+use crate::core::alignment;
+use crate::core::text::{Hit, LineHeight, Shaping, Text};
+use crate::core::{Font, Pixels, Point, Size};
+use crate::text::{self, FontSystem};
+
+use std::fmt;
+use std::sync::{self, Arc};
+
+#[derive(Clone, PartialEq, Default)]
+pub struct Paragraph(Arc<Internal>);
+
+struct Internal {
+    buffer: cosmic_text::Buffer,
+    content: String, // TODO: Reuse from `buffer` (?)
+    font: Font,
+    shaping: Shaping,
+    horizontal_alignment: alignment::Horizontal,
+    vertical_alignment: alignment::Vertical,
+    bounds: Size,
+    min_bounds: Size,
+}
+
+impl Paragraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_text(text: Text<'_, Font>, font_system: &FontSystem) -> Self {
+        let mut font_system = font_system.write();
+
+        let mut buffer = cosmic_text::Buffer::new(
+            &mut font_system,
+            cosmic_text::Metrics::new(
+                text.size.into(),
+                text.line_height.to_absolute(text.size).into(),
+            ),
+        );
+
+        buffer.set_size(
+            &mut font_system,
+            text.bounds.width,
+            text.bounds.height,
+        );
+
+        buffer.set_text(
+            &mut font_system,
+            text.content,
+            text::to_attributes(text.font),
+            text::to_shaping(text.shaping),
+        );
+
+        let min_bounds = text::measure(&buffer);
+
+        Self(Arc::new(Internal {
+            buffer,
+            content: text.content.to_owned(),
+            font: text.font,
+            horizontal_alignment: text.horizontal_alignment,
+            vertical_alignment: text.vertical_alignment,
+            shaping: text.shaping,
+            bounds: text.bounds,
+            min_bounds,
+        }))
+    }
+
+    pub fn buffer(&self) -> &cosmic_text::Buffer {
+        &self.0.buffer
+    }
+
+    pub fn downgrade(&self) -> Weak {
+        Weak {
+            raw: Arc::downgrade(&self.0),
+            min_bounds: self.0.min_bounds,
+            horizontal_alignment: self.0.horizontal_alignment,
+            vertical_alignment: self.0.vertical_alignment,
+        }
+    }
+
+    pub fn resize(&mut self, new_bounds: Size, font_system: &FontSystem) {
+        if let Some(internal) = Arc::get_mut(&mut self.0) {
+            // If there is no strong reference holding on to the paragraph, we
+            // resize the buffer in-place
+            internal.buffer.set_size(
```

Comment:
Good catch! We definitely want `Arc::into_inner`.

`Weak` references should only be held by the `Primitive` tree, and they will only be invalidated right before creating a new one.

This makes the `Primitive` trees much cheaper to clone for incremental rendering and, since a `paragraph::Weak` holds its original bounds and alignment, it is still possible to figure out its damage region.

Widgets should never be handling a `Weak` primitive directly.

Reply:
`try_unwrap` needs ownership, so we need to obtain it from a mutable reference.

I have implemented an unsafe version with `MaybeUninit` in 935c722a278071d3bcd0bf499a0ef8f687ed51ad, and another safe version with `Option` in 51e69d7040c943aad8453bc03031c75b4cb302bb.

I think the `MaybeUninit` is sound because the memory is only uninitialized while we have unique access to it, but I don't feel very comfortable with unsafe yet.

Let's go with the `Option` approach; I don't think the overhead will be noticeable in practice.